pkg/querier/distributor_queryable_streaming.go

```diff
@@ -0,0 +1,164 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querier
+
+import (
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+
+	"github.com/grafana/mimir/pkg/ingester/client"
+	"github.com/grafana/mimir/pkg/storage/series"
+	"github.com/grafana/mimir/pkg/util/chunkcompat"
+)
+
+type streamingChunkSeries struct {
+	labels            labels.Labels
+	chunkIteratorFunc chunkIteratorFunc
+	mint, maxt        int64
+	sources           []StreamingSeriesSource
+}
+
+func (s *streamingChunkSeries) Labels() labels.Labels {
+	return s.labels
+}
+
+func (s *streamingChunkSeries) Iterator(_ chunkenc.Iterator) chunkenc.Iterator {
+	var rawChunks []client.Chunk // TODO: guess a size for this?
+
+	for _, source := range s.sources {
+		c, err := source.Streamer.GetChunks(source.SeriesIndex)
+
+		if err != nil {
+			return series.NewErrIterator(err)
+		}
+
+		rawChunks = client.AccumulateChunks(rawChunks, c)
+	}
+
+	chunks, err := chunkcompat.FromChunks(s.labels, rawChunks)
+	if err != nil {
+		return series.NewErrIterator(err)
+	}
+
+	return s.chunkIteratorFunc(chunks, model.Time(s.mint), model.Time(s.maxt))
+}
+
+type SeriesStreamer struct {
```

Comment:
[nit] When I read "streamer" it looks like a server-side component to me (something that streams). Here it looks more like a "stream reader". Maybe we can rename it to `SeriesChunksStreamReader`?

Reply:
Makes sense, I've renamed it in 49a8b3e